### ISSUE TITLE
Verifying GitHub specific claims

### ISSUE THREAD

**Issue by jku** [sigstore/sigstore-python#157]
It would be very useful if it was possible to verify the GitHub specific claims in the certificate that we produce when using GitHub Actions. This is because the claims allow verifying that:
* Signing happened on GitHub Actions builder...
* while running the specific workflow...
* using the specified commit SHA..
* of the specified GitHub project

**This completely defines the build process in an auditable way**.

This issue is related to #106 and #124 but I figured it would make sense to have a specific issue as I have prototype code... 

I've said this before but I'll restate: this GitHub-specific functionality does not have to live in sigstore-python but we should make sure it's somehow possible.

**Comment by jku** [comment]
See https://github.com/jku/sigstore-python/commits/github-verify for a prototype. It's a bit rough but I'll open a draft PR one of these days...

In my branch I've included the certificate in the VerificationSuccess result: and this would also make it possible to implement this kind of functionality outside of the sigstore-python codebase, assuming the Verifier is public API.

Usage:
```
python -m sigstore verify-github \
    --repository sigstore/sigstore-python \
    --workflow .github/workflows/release.yml \
    --tag v0.5.1 \
    --sha 05b7ab4de650271873469ef087273c28d1cac4d7 \
    sigstore-0.5.1-py3-none-any.whl
```
`tag` and `sha` are optional, others are required.

**Comment by di** [comment]
Agreed that we should allow users to verify any information that's available and not artificially prevent them from doing so. I do have concerns about this becoming unwieldy given the number of claims/args we'd need to support (multiplied by all the different OIDC providers) so I'd definitely want to think about this in the context of https://github.com/sigstore/sigstore-python/issues/124 and with a thought towards how we can make this easily generalizable (i.e., not require changes to the CLI if we introduce a new supported OIDC provider with different claims).

**Comment by jku** [comment]
> I do have concerns about this becoming unwieldy given the number of claims/args we'd need to support (multiplied by all the different OIDC providers) so I'd definitely want to think about this in the context of #124 and with a thought towards how we can make this easily generalizable

Yeah, it is a very valid concern and the main reason for not just creating a PR... I've played with the configuration format idea as well and see two problems with it:
* To solve the problem in this issue, you still need to provide an _easy and safe way to define what claims should be verified for the GitHub case_ out-of-band (as in, the config file can't be downloaded with the certs and signatures). I believe you still end up creating a script that produces a known good configuration file for GitHub (and then another script for another provider). If you end up creating a script, why not just call the Verifier API from the script directly?
* the configuration easily becomes quite a bit more verbose and difficult to understand than my CLI example is and even then it could be difficult to make it generic enough to cover everything that could be inserted into cert extensions

A potential path forward is to make the Verifier API generally useful, and implement the provider specific tools as separate from the main `sigstore` module -- in other modules or outside the project even

**Comment by jku** [comment]
So I've come back to this after a bit of a break and I think this is still the state of things:
* It's currently not possible to verify anything useful about the claims that GitHub makes in their certificate (the only thing sigstore-python can verify is that the issuer is GitHub): If I want to verify _anything_ else I need to start parsing the certificate file myself
* The idea of adding GitHub-specific functionality to sigstore-python has received appropriate levels of scepticism here :) I admit I agree with this criticism
* sigstore-python API currently does not make it easy to create this sort of functionality elsewhere

I think the last item could be fairly easy to fix. I'll open a draft PR to get some opinions on this approach.

**Comment by di** [comment]
Yes, I think we're blocked on some movement on https://github.com/sigstore/cosign/issues/1964 so we can resolve https://github.com/sigstore/sigstore-python/issues/108 and have a way to verify arbitrary extensions (without needing to provide some GitHub-specific mechanism).

**Comment by jku** [comment]
The --cert-email/--cert-identity issue is clearly solvable... but I'm not sure exposing arbitrary extensions can lead to acceptable UX (whether the extensions are exposed to user via a command line interface or a configuration file that the verifier writes)? If it can be successfully done then that would certainly solve the problem.

Some issues I've seen with exposing extensions (examples for GitHub obviously):
* readability is pretty awful: `--check-extension 1.3.6.1.4.1.57264.1.3 bff2e635da74627b62b31efd746f533bf97801ed` -- it's quite hard to see if that is the correct check or not
* logical input from user is items like `repository` and `tag`, but what certificate actually contains is e.g. `URI:https://github.com/sigstore/sigstore-python/.github/workflows/release.yml@refs/tags/v0.6.3` -- requiring that from a user seems a bit much
* it's beneficial to be able to choose how to validate based in input: e.g. if tag is not given as input then checking the SAN prefix is still useful as it fully verifies the project -- this is hard to express without code

**Comment by haydentherapper** [comment]
FYI this was added in https://github.com/sigstore/cosign/pull/1626 for Cosign. We went with individual flags for each GitHub claim.

**Comment by haydentherapper** [comment]
I think the UX isn't ideal, and that the long-term solution would be a policy file. For one-off verifications, I think flags are fine.

**Comment by woodruffw** [comment]
I'm going to include the backing work for this in #250.

**Comment by woodruffw** [comment]
The API is done here, so I'm closing this in favor of #332 for tracking the CLI support.
